notebooks/work-in-progress/ferc1-transform-refactor/ferc1_transform_refactor.py

```diff
@@ -0,0 +1,429 @@
+    def from_id(cls, table_id: str) -> "TableTransformParams":
+        """A factory method that looks up transform parameters based on table_id."""
+        return cls(**TRANSFORM_PARAMS[table_id])
+
+
+################################################################################
+# Column, MultiColumn, and Table Transform Functions
+################################################################################
+class ColumnTransformFn(Protocol):
+    """Callback protocol defining a per-column transformation function."""
+
+    def __call__(self, col: pd.Series, params: TransformParams) -> pd.Series:
+        """Create a callable."""
+        ...
+
+
+class TableTransformFn(Protocol):
+    """Callback protocol defining a per-table transformation function."""
+
+    def __call__(self, df: pd.DataFrame, params: TransformParams) -> pd.DataFrame:
+        """Create a callable."""
+        ...
+
+
+def multicol_transform_fn_factory(
+    col_fn: ColumnTransformFn,
+    drop=False,
+) -> TableTransformFn:
+    """A factory for creating a table transformation from a column transformation."""
+
+    def tab_fn(df: pd.DataFrame, params: MultiColumnTransformParams) -> pd.DataFrame:
+        drop_col: bool = drop
+        for col_name in params:
+            if col_name in df.columns:
+                new_col = col_fn(col=df[col_name], params=params[col_name])
+                if drop_col:
+                    df = df.drop(columns=col_name)
+                df = pd.concat([df, new_col], axis="columns")
+            else:
+                logger.warning(
+                    f"Expected column {col_name} not found in dataframe during table "
+                    "transform."
+                )
+        return df
+
+    return tab_fn
+
+
+def convert_units_col(col: pd.Series, params: UnitConversion) -> pd.Series:
+    """Convert the units of and appropriately rename a column."""
+    new_name = re.sub(pattern=params.pattern, repl=params.repl, string=col.name)
+    # only apply the unit conversion if the column name matched the pattern
+    if new_name == col.name:
+        logger.warning(
+            f"{col.name} did not match the unit rename pattern. Check for typos "
+            "and make sure you're applying the conversion to an appropriate column."
+        )
+    col = (params.multiplier * col) + params.adder
+    col.name = new_name
+    return col
+
+
+convert_units = multicol_transform_fn_factory(convert_units_col, drop=True)
+
+
+def categorize_strings_col(col: pd.Series, params: StringCategories) -> pd.Series:
+    """Impose a controlled vocabulary on freeform string column."""
+    return col.map(params.mapping)
+
+
+categorize_strings = multicol_transform_fn_factory(categorize_strings_col, drop=True)
+
+
+################################################################################
+# TableTransformer classes
+################################################################################
+class AbstractTableTransformer(ABC):
```

Comment:
`AbstractTableTransformer` could provide a forensic class method that returns a representation of all TransformFn and TransformParams so users can easily view all of the transformations a table goes through instead of roaming around in the code.

Reply:
This would make it less important to have the parameters stored in the module.

I'm also struggling with how to organize the several different kinds of transformation procedures:
* standalone functions, which are available outside of the `TableTransformer` context because they're useful elsewhere too.
* factory generated stand-alone multi-column transform functions that are derived from a standalone column transform function.
* transform methods that apply only within a given dataset, like the `ferc1`
* transform methods that apply only to a particular table, like `plants_steam_ferc1`
* For transform methods, how to avoid duplication when defining a multi-column transform derived from a single-column transform.

It would be nice if they could all play by the same rules, but the need to have `self` or `cls` as an argument inside the classes seems like it changes the interface compared to a function living free in the outside world.